crates/iceberg-catalog/src/service/contract_verification.rs

```diff
@@ -0,0 +1,265 @@
+///                     .build()
+///                     .into(),
+///             })
+///         }
+///     }
+/// ```
+#[async_trait]
+pub trait ContractVerification: Debug {
+    fn name(&self) -> &'static str;
+
+    async fn check(
+        &self,
+        table_updates: &[TableUpdate],
+        table_ident_uuid: TableIdentUuid,
+        current_metadata: &TableMetadata,
+    ) -> Result<ContractVerificationOutcome, ErrorModel>;
+
+    async fn check_drop(
+        &self,
+        table_ident_uuid: TableIdentUuid,
+    ) -> Result<ContractVerificationOutcome, ErrorModel>;
+
+    async fn check_rename(
+        &self,
+        source: TableIdentUuid,
+        destination: &TableIdent,
+    ) -> Result<ContractVerificationOutcome, ErrorModel>;
+}
+
+#[derive(Debug)]
+pub enum ContractVerificationOutcome {
+    Clear {},
+    Violation { error_model: ErrorModel },
+}
+
+impl ContractVerificationOutcome {
+    /// Converts `self` into a `Result<(), ErrorModel>`.
+    ///
+    /// When using `ContractVerificationOutcome`, we are presented with a
+    /// `Result<ContractVerificationOutcome, ErrorModel>`
+    /// where the outer `ErrorModel` indicates that a checker failed, that would indicate a problem
+    /// with the checker itself and may be returned as an internal server error. This function here
+    /// offers convenience to go from the `ContractVerificationOutcome` to a `Result<(), ErrorModel>`
+    /// which can be short-circuited using the `?` operator in the handler.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    ///     use iceberg_catalog::service::contract_verification::ContractVerificationOutcome;
+    ///     use iceberg_ext::catalog::rest::ErrorModel;
+    ///
+    ///     fn my_handler() -> Result<(), ErrorModel> {
+    ///         let result: Result<ContractVerificationOutcome, ErrorModel> = Ok(ContractVerificationOutcome::Clear {});
+    ///         result?.into_result()?;
+    ///         Ok(())
+    ///     }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// - extracts `error_model` from `ContractVerificationOutcome::Block` and returns it as an `Err` for
+    ///   convenience.
+    pub fn into_result(self) -> Result<(), ErrorModel> {
+        match self {
+            ContractVerificationOutcome::Clear {} => Ok(()),
+            ContractVerificationOutcome::Violation { error_model } => Err(error_model),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ContractVerifiers {
+    checkers: Vec<Arc<dyn ContractVerification + Sync + Send>>,
+}
+
+impl ContractVerifiers {
+    #[must_use]
+    pub fn new(checkers: Vec<Arc<dyn ContractVerification + Sync + Send>>) -> Self {
+        Self { checkers }
+    }
+}
+
+#[async_trait]
+impl ContractVerification for ContractVerifiers {
+    fn name(&self) -> &'static str {
+        "ContractVerifiers"
+    }
+
+    async fn check(
+        &self,
+        table_updates: &[TableUpdate],
+        table_ident_uuid: TableIdentUuid,
```

Comment:
Now that we send the ` current_metadata`, we might want to skip the table_ident_uuid.
The table-uuid is already part of `TableMetadata` and might change during updates.
If we  specify it here additionally, we should rename it to current_table_ident_uuid to make sure it is the UUID before updates.
I think removing it is fine though.

Reply:
thanks, missed it in table-metadata, removed!